File: cpp/union_find.cpp
/*
Union-find (disjoint-set union, DSU) maintains a collection of disjoint sets under two operations:

* find(x): return the representative (root) of the set containing x.
* union(x, y): merge the sets containing x and y.

Time complexity: O(alpha(n)) per operation with path compression and union by rank,
where alpha is the inverse Ackermann function (effectively constant for practical purposes).
*/

#include <cassert>
#include <iostream>
#include <set>

class UnionFind {
  public:
    UnionFind* parent;
    int rank;

    UnionFind() : parent(this), rank(0) {}

    virtual void merge(UnionFind* other) {
        // Override with desired functionality
    }

    UnionFind* find() {
        if (parent == this) { return this; }
        parent = parent->find();
        return parent;
    }

    UnionFind* union_with(UnionFind* other) {
        UnionFind* x = this->find();
        UnionFind* y = other->find();
        if (x == y) { return x; }
        if (x->rank < y->rank) {
            x->parent = y;
            y->merge(x);
            return y;
        }
        if (x->rank > y->rank) {
            y->parent = x;
            x->merge(y);
            return x;
        }
        x->parent = y;
        y->merge(x);
        y->rank++;
        return y;
    }
};

class Test : public UnionFind {
  public:
    int size;

    Test() : UnionFind(), size(1) {}

    void merge(UnionFind* other) override {
        Test* other_test = static_cast<Test*>(other);
        this->size += other_test->size;
    }
};

void test_main() {
    Test* a = new Test();
    Test* b = new Test();
    Test* c = new Test();
    Test* d = static_cast<Test*>(a->union_with(b));
    Test* e = static_cast<Test*>(d->union_with(c));
    assert(static_cast<Test*>(e->find())->size == 3);
    assert(static_cast<Test*>(a->find())->size == 3);

    delete a;
    delete b;
    delete c;
}

// Don't write tests below during competition.

void test_single_element() {
    Test* a = new Test();
    assert(a->find() == a);
    assert(a->size == 1);
    delete a;
}

void test_union_same_set() {
    Test* a = new Test();
    Test* b = new Test();
    a->union_with(b);
    // Unioning again should be safe
    Test* root = static_cast<Test*>(a->union_with(b));
    assert(a->find() == b->find());
    assert(root->size == 2);
    delete a;
    delete b;
}

void test_multiple_unions() {
    Test* nodes[10];
    for (int i = 0; i < 10; i++) { nodes[i] = new Test(); }

    // Chain union: 0-1-2-3-4-5-6-7-8-9
    for (int i = 0; i < 9; i++) { nodes[i]->union_with(nodes[i + 1]); }

    // All should have same root
    UnionFind* root = nodes[0]->find();
    for (int i = 0; i < 10; i++) { assert(nodes[i]->find() == root); }

    assert(static_cast<Test*>(root)->size == 10);

    for (int i = 0; i < 10; i++) { delete nodes[i]; }
}

void test_union_order_independence() {
    // Test that union order doesn't affect final result
    Test* a1 = new Test();
    Test* b1 = new Test();
    Test* c1 = new Test();
    a1->union_with(b1)->union_with(c1);
    Test* root1 = static_cast<Test*>(a1->find());

    Test* a2 = new Test();
    Test* b2 = new Test();
    Test* c2 = new Test();
    c2->union_with(b2)->union_with(a2);
    Test* root2 = static_cast<Test*>(a2->find());

    assert(root1->size == 3);
    assert(root2->size == 3);

    delete a1;
    delete b1;
    delete c1;
    delete a2;
    delete b2;
    delete c2;
}

void test_disconnected_sets() {
    // Create two separate sets
    Test* a = new Test();
    Test* b = new Test();
    Test* c = new Test();
    Test* d = new Test();

    a->union_with(b);
    c->union_with(d);

    assert(a->find() == b->find());
    assert(c->find() == d->find());
    assert(a->find() != c->find());

    assert(static_cast<Test*>(a->find())->size == 2);
    assert(static_cast<Test*>(c->find())->size == 2);

    delete a;
    delete b;
    delete c;
    delete d;
}

void test_large_set() {
    // Create a large union-find structure
    Test* nodes[100];
    for (int i = 0; i < 100; i++) { nodes[i] = new Test(); }

    // Union in pairs
    for (int i = 0; i < 100; i += 2) { nodes[i]->union_with(nodes[i + 1]); }

    // Now we have 50 sets of size 2
    std::set<UnionFind*> roots;
    for (int i = 0; i < 100; i++) { roots.insert(nodes[i]->find()); }
    assert(roots.size() == 50);

    // Union all pairs together
    for (int i = 0; i < 100; i += 4) {
        if (i + 2 < 100) { nodes[i]->union_with(nodes[i + 2]); }
    }

    // Now we have 25 sets of size 4
    roots.clear();
    for (int i = 0; i < 100; i++) { roots.insert(nodes[i]->find()); }
    assert(roots.size() == 25);

    for (int i = 0; i < 100; i++) { delete nodes[i]; }
}

int main() {
    test_single_element();
    test_union_same_set();
    test_multiple_unions();
    test_union_order_independence();
    test_disconnected_sets();
    test_large_set();
    test_main();
    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: python/union_find.py
"""
Union-find (disjoint-set union, DSU) maintains a collection of disjoint sets under two operations:

* find(x): return the representative (root) of the set containing x.
* union(x, y): merge the sets containing x and y.

Time complexity: O(alpha(n)) per operation with path compression and union by rank,
where alpha is the inverse Ackermann function (effectively constant for practical purposes).
"""

from __future__ import annotations

# Don't use annotations during contest
from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:
    from typing_extensions import Self


class UnionFind:
    def __init__(self) -> None:
        self.parent = self
        self.rank = 0

    def merge(self, other: Self) -> None:
        """Override to define custom merge behavior when sets are united."""

    def find(self) -> Self:
        """Return root of this set with path compression."""
        if self.parent == self:
            return self
        self.parent = self.parent.find()
        return cast("Self", self.parent)

    def union(self, other: Self) -> Self:
        """Unite sets containing self and other. Returns the new root."""
        x = self.find()
        y = other.find()
        if x is y:
            return x
        if x.rank < y.rank:
            x.parent = y
            y.merge(x)
            return y
        if x.rank > y.rank:
            y.parent = x
            x.merge(y)
            return x
        x.parent = y
        y.merge(x)
        y.rank += 1
        return y


class Test(UnionFind):
    """Better to modify copy of UnionFind class and avoid having to type cast everywhere."""

    def __init__(self) -> None:
        super().__init__()
        self.size = 1

    def merge(self, other: Self) -> None:
        assert isinstance(other, Test)
        self.size += other.size


def test_main() -> None:
    a, b, c = Test(), Test(), Test()
    d = a.union(b)
    e = d.union(c)
    assert e.find().size == 3
    assert a.find().size == 3


# Don't write tests below during competition.


def test_single_element() -> None:
    a = Test()
    assert a.find() is a
    assert a.size == 1


def test_union_same_set() -> None:
    a = Test()
    b = Test()
    a.union(b)
    # Unioning again should be safe
    root = a.union(b)
    assert a.find() is b.find()
    assert root.size == 2


def test_multiple_unions() -> None:
    nodes = [Test() for _ in range(10)]
    # Chain union: 0-1-2-3-4-5-6-7-8-9
    for i in range(9):
        nodes[i].union(nodes[i + 1])

    # All should have same root
    root = nodes[0].find()
    for node in nodes:
        assert node.find() is root

    assert root.size == 10


def test_union_order_independence() -> None:
    # Test that union order doesn't affect final result
    a1, b1, c1 = Test(), Test(), Test()
    a1.union(b1).union(c1)
    root1 = a1.find()

    a2, b2, c2 = Test(), Test(), Test()
    c2.union(b2).union(a2)
    root2 = a2.find()

    assert root1.size == root2.size == 3


def test_disconnected_sets() -> None:
    # Create two separate sets
    a, b = Test(), Test()
    c, d = Test(), Test()

    a.union(b)
    c.union(d)

    assert a.find() is b.find()
    assert c.find() is d.find()
    assert a.find() is not c.find()

    assert a.find().size == 2
    assert c.find().size == 2


def test_large_set() -> None:
    # Create a large union-find structure
    nodes = [Test() for _ in range(100)]

    # Union in pairs
    for i in range(0, 100, 2):
        nodes[i].union(nodes[i + 1])

    # Now we have 50 sets of size 2
    roots = set()
    for node in nodes:
        roots.add(id(node.find()))
    assert len(roots) == 50

    # Union all pairs together
    for i in range(0, 100, 4):
        if i + 2 < 100:
            nodes[i].union(nodes[i + 2])

    # Now we have 25 sets of size 4
    roots = set()
    for node in nodes:
        roots.add(id(node.find()))
    assert len(roots) == 25


def main() -> None:
    test_single_element()
    test_union_same_set()
    test_multiple_unions()
    test_union_order_independence()
    test_disconnected_sets()
    test_large_set()
    test_main()


if __name__ == "__main__":
    main()
